crates/subspace-farmer/src/single_disk_farm.rs

```diff
@@ -122,10 +158,291 @@ impl SingleDiskFarmMetadata {
+///
+/// It primarily constraints some of the disk access concurrency to achieve higher performance by
+/// avoiding unnecessary random disk access and preferring sequential reads/writes whenever possible
+/// instead of doing a large amount of random I/O that is bad for HDDs (intended storage medium for
+/// plots).
+pub struct SingleDiskFarm {
+    single_plot_farms: Vec<SinglePlotFarm>,
+    archiving: Option<Archiving>,
+}
+
+impl SingleDiskFarm {
+    /// Creates single disk farm with user-provided total plot size
+    pub async fn new<RC, PF>(options: SingleDiskFarmOptions<RC, PF>) -> anyhow::Result<Self>
+    where
+        RC: RpcClient,
+        PF: PlotFactory,
+    {
+        let SingleDiskFarmOptions {
+            plot_directory,
+            metadata_directory,
+            usable_plotting_space,
+            farmer_protocol_info,
+            plot_factory,
+            archiving_client,
+            farming_client,
+            reward_address,
+            bootstrap_nodes,
+            listen_on,
+            enable_dsn_archiving,
+            enable_dsn_sync,
+            enable_farming,
+        } = options;
+
+        let plot_sizes = get_plot_sizes(usable_plotting_space, farmer_protocol_info.max_plot_size);
+
+        let single_disk_farm_metadata =
+            match SingleDiskFarmMetadata::load_from(&metadata_directory)? {
+                Some(single_disk_farm_metadata) => {
+                    if usable_plotting_space != single_disk_farm_metadata.usable_plotting_space() {
+                        error!(
+                            id = %single_disk_farm_metadata.id(),
+                            plot_directory = %plot_directory.display(),
+                            metadata_directory = %metadata_directory.display(),
+                            "Usable plotting space {} is different from {} when farm was created, \
+                            resizing isn't supported yet",
+                            usable_plotting_space,
+                            single_disk_farm_metadata.usable_plotting_space(),
+                        );
+
+                        return Err(anyhow!("Can't resize farm after creation"));
+                    }
+
+                    if &farmer_protocol_info.genesis_hash
+                        != single_disk_farm_metadata.genesis_hash()
+                    {
+                        error!(
+                            id = %single_disk_farm_metadata.id(),
+                            "Genesis hash {} is different from {} when farm was created, is is not \
+                            possible to use farm on a different chain",
+                            hex::encode(farmer_protocol_info.genesis_hash),
+                            hex::encode(single_disk_farm_metadata.genesis_hash()),
+                        );
+
+                        return Err(anyhow!("Wrong chain (genesis hash)"));
+                    }
+                    single_disk_farm_metadata
+                }
```

Comment:
I bet it would look simpler with match guards:
```rust
                   Some(single_disk_farm_metadata) if usable_plotting_space != single_disk_farm_metadata.usable_plotting_space() => {
                        error!(
                            id = %single_disk_farm_metadata.id(),
                            plot_directory = %plot_directory.display(),
                            metadata_directory = %metadata_directory.display(),
                            "Usable plotting space {} is different from {} when farm was created, \
                            resizing isn't supported yet",
                            usable_plotting_space,
                            single_disk_farm_metadata.usable_plotting_space(),
                        );

                        return Err(anyhow!("Can't resize farm after creation"));
                   }
...
```

Reply:
Super long lines like that do not really look simple, also in IDE you can collapse logical blocks, with this flattening you disable that ability for no reason.

It achieves the same result, sure, but I think it is better to write the code in a way that makes intentions obvious. For instance if we have a number on `Some()` and wanted to do 3 different things with a lot of code based on whether that number is negative, positive or zero then the suggested approach is reasonable. Here we are checking a few error cases and return the value itself, I don't see how flattening it makes things easier to understand this.